File: Baekjoon/99mini/1660.js
const readline = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];
readline
  .on("line", function (line) {
    input = line.split(" ").map((el) => parseInt(el));
  })
  .on("close", function () {
    const n = parseInt(input[0]);
    const balls = [];

    let b = 0;
    let i = 1;

    while (b < n) {
      b += Math.floor((i * (i + 1)) / 2); // 정사면체에 필요한 포탄의 개수. 1, 4, 10, 20
      balls.push(b);
      i += 1;
    }

    let dp = Array.from({ length: n + 1 }).fill(Number.MAX_SAFE_INTEGER);

    for (let i = 1; i < n + 1; i++) {
      for (const b of balls) {
        if (b === i) {
          dp[i] = 1;
        }
        if (b >= i) {
          break;
        }
        dp[i] = Math.min(dp[i], 1 + dp[i - b]);
      }
    }

    console.log(dp[n]);
    process.exit();
  });

File: Baekjoon/99mini/1660.py
'''
작성자: 99mini
문제 링크: https://www.acmicpc.net/problem/1660
카테고리
- 다이나믹 프로그래밍

문제 해설
정사면체의 누적합을 배열로 관리
'''

import sys

input = sys.stdin.readline

MAX = sys.maxsize

n = int(input())

balls = []
b = 0
i = 1
while b < n:
    b += (i * (i + 1)) // 2 # 정사면체에 필요한 포탄의 개수. 1, 4, 10, 20
    balls.append(b)
    i+=1

dp = [MAX]*(n+1)

for i in range(1, n+1):
    for b in balls:
        if b == i:
            dp[i] = 1
        if b >= i:
            break
        dp[i] = min(dp[i], 1 + dp[i - b])
print(dp[n])
